### ISSUE TITLE
Period codes

### ISSUE THREAD

**Issue by gmoger-jisc** [jiscdev/analytics-udd#358]
https://github.com/jiscdev/analytics-udd/blob/master/udd/period.md#period_code

It would be useful to clarify our approach here and perhaps implement a list of valid values or reserved values:

- ACADYR
- TERM

This will help us identify the core elements that make up the academic year. 

Thoughts?

**Comment by arc12** [comment]
it is a bit a a hybrid, I agree, but I think the essential requirement is that PERIOD provides a lookup which we use for ACADEMIC_YEAR, MOD_ACADEMIC_YEAR, MOD_PERIOD, and COMMENCEMENT_PERIOD.

We rely on this linkage.

If the requirement is to identify the elements, maybe this could be achieved with a separate period-type attribute?

**Comment by gmoger-jisc** [comment]
Yes I think a separate field would make this easier but I was trying to avoid that if possible.

My requirement for this is to ensure that MOD_PERIOD reflects a term in the period entity but if we wanted to understand the terms within an academic year then it would be difficult to do this from the period entity alone.

**Comment by gmoger-jisc** [comment]
Suggesting to remove period code, which is being used as the institutions id and contains this purpose:

> Where the period record states the dates for the institution's academic year, PERIOD_CODE must use the fixed string "ACADYR". For other periods, PERIOD_CODE uses institution-defined code values. Each different code value in module_instance.MOD_PERIOD or course_instance.COMMENCEMENT_PERIOD should have a matching code value in PERIOD_CODE. Where the same code values are used across different academic years, the specific period record can be identified by reference to the PERIOD_CODE together with the ACADEMIC_YEAR property; these properties together form the primary key.

PERIOD_ID is the primary and can be institutions own code; a separate field feels like duplicates.

To accommodate our need to distinguish the type of period as being either a term or an academic year, a new field has been added with valid values.  These can be extended if needed but these 2 should represent the core of time slices in UK education.

@arc12 how would this effect your existing usage?

**Comment by arc12** [comment]
This will break our current use.

The way we've understood this is that the PERIOD_CODE represents things like "michaelmas term" which repeats every academic year, so that this is part of a uniqueness constraint with ACADEMIC_YEAR.
Examples of PERIOD_CODE include module periods spanning 2 or 3 terms, so we should view this data as being something which can differ from a list of standard academic periods.

We view PERIOD_ID as a surrogate PK. We dont use this (although we do populate in the ETL), and I understood it could be generated by the Data Hub.

so: a vote for the status quo

**Comment by gmoger-jisc** [comment]
I think you're right and it makes more sense to me now thinking about modules spanning terms.

In which case this might just need to be made more clear in the spec.

To add more context to our use case, we need to report on attendance/engagement for UKVI, which requires a per term review.  Ideally there would be some way to capture this in the period entity but that becomes difficult if there are multiple overlapping terms defined.
